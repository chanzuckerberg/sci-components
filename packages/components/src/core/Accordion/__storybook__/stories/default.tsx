import { Args } from "@storybook/react";
import RawAccordion, {
  AccordionDetails,
  AccordionHeader,
} from "src/core/Accordion";
import { EXTRA_SHORT_LOREM_IPSUM } from "src/common/storybook/loremIpsum";

export const Accordion = (props: Args): JSX.Element => {
  const { id, subtitle, useDivider, togglePosition } = props;
  return (
    <RawAccordion
      id={id}
      useDivider={useDivider}
      togglePosition={togglePosition}
      {...props}
    >
      <AccordionHeader id={`${id}-header`} subtitle={subtitle}>
        Accordion Header
      </AccordionHeader>
      <AccordionDetails>{EXTRA_SHORT_LOREM_IPSUM}</AccordionDetails>
    </RawAccordion>
  );
};
