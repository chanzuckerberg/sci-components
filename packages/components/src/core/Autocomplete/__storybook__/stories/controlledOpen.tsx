import { Args } from "@storybook/types";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { useState } from "react";
import RawAutocomplete from "src/core/Autocomplete";
import { AUTOCOMPLETE_WRAPPER_STYLES } from "../constants";

export const ControlledOpenDemo = <T extends DefaultAutocompleteOption>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
  } = props;

  const [open, setOpen] = useState(true);

  return (
    <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
      <RawAutocomplete
        onClickAway={() => {
          setOpen(false);
        }}
        onClick={() => {
          if (open) {
            setOpen(false);
          } else {
            setOpen(true);
          }
        }}
        open={open}
        id="autocomplete-demo"
        disableCloseOnSelect={multiple}
        label={label}
        multiple={multiple}
        options={options}
        getOptionDisabled={(option: T) => {
          return option.name === "Type: feature request";
        }}
        {...props}
      />
    </div>
  );
};
