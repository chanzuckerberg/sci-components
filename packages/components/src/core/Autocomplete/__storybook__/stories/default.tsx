import { Args } from "@storybook/react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS";
import { SyntheticEvent, useEffect, useState } from "react";
import RawAutocomplete, {
  AutocompleteMultiColumnValue,
  SDSAutocompleteValue,
} from "src/core/Autocomplete";
import TagFilter from "src/core/TagFilter";
import { AutocompleteValue } from "@mui/base";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";

export const Autocomplete = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    label,
    multiple,
    options = AUTOCOMPLETE_SINGLE_COLUMN_OPTIONS,
    search,
    value: propValue,
    keepSearchOnSelect,
  } = props;

  type DisableClearable = false;
  type FreeSolo = false;

  const isControlled = propValue !== undefined;
  const [value, setValue] = useState<
    SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  >(
    (multiple ? [] : null) as SDSAutocompleteValue<
      T,
      Multiple,
      DisableClearable,
      FreeSolo
    >
  );

  const [selection, setSelection] = useState<string[]>([]);

  useEffect(() => {
    if (isControlled) {
      setValue(propValue);
    }
  }, [isControlled, propValue]);

  useEffect(() => {
    setSelection([]);
    setValue(
      (multiple ? [] : null) as SDSAutocompleteValue<
        T,
        Multiple,
        DisableClearable,
        FreeSolo
      >
    );
  }, [multiple, options]);

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <RawAutocomplete
          id="autocomplete-demo"
          disableCloseOnSelect={multiple}
          label={label}
          multiple={multiple}
          onChange={handleChange}
          keepSearchOnSelect={keepSearchOnSelect}
          options={options}
          search={search}
          value={value}
          getOptionDisabled={(option: T) => {
            return option.name === "Type: feature request";
          }}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleChange(
    _: SyntheticEvent<Element, Event>,
    newValue: SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
  ) {
    setValue(newValue);
    const newSelection: string[] = [];

    if (
      (newValue as T)?.name ||
      (Array.isArray(newValue) && (newValue as T[])[0].name)
    ) {
      if (multiple) {
        setSelection(
          Array.isArray(newValue) ? newValue?.map((item) => item.name) : []
        );
      } else {
        if (newValue && !Array.isArray(newValue) && newValue.name) {
          setSelection([(newValue as T).name]);
        }
      }
    } else {
      if (multiple) {
        if (newValue) {
          Object.values(newValue).forEach((items) => {
            (items as T[])?.map(({ name }: { name: string }) => {
              newSelection.push(name);
            });
          });
          setSelection(newSelection);
        }
      } else {
        if (newValue) {
          Object.values(newValue).forEach((item) => {
            newSelection.push((item as T).name);
          });
        }
        setSelection(newSelection);
      }
    }
  }

  // eslint-disable-next-line sonarjs/cognitive-complexity
  function handleTagDeleteMultiColumn(tag: string) {
    if (multiple) {
      if (value) {
        Object.entries(value).forEach(([key, values]) => {
          const index = (values as T[])?.findIndex(
            (item: T) => item.name === tag
          );
          if (index > -1) {
            const newValue = [...(values as T[])];
            newValue.splice(index, 1);
            setValue((prev) => {
              return {
                ...prev,
                [key]: newValue as AutocompleteValue<
                  T,
                  Multiple,
                  DisableClearable,
                  FreeSolo
                >,
              };
            });

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    } else {
      if (value) {
        Object.entries(value).forEach(([key, val]) => {
          if ((val as T).name === tag) {
            const finalValue = { ...value } as AutocompleteMultiColumnValue<
              T,
              Multiple,
              DisableClearable,
              FreeSolo
            >;
            if (finalValue) delete finalValue[key];
            setValue(finalValue);

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    }
  }

  function handleTagDelete(tag: string) {
    if (
      (value && (value as T).name) ||
      (Array.isArray(value) && (value as T[])[0].name)
    ) {
      if (multiple && Array.isArray(value)) {
        const index = value?.findIndex((item) => item.name === tag);
        const newValue = [...value];
        newValue.splice(index, 1);
        setValue(
          newValue as SDSAutocompleteValue<
            T,
            Multiple,
            DisableClearable,
            FreeSolo
          >
        );

        const newSelection = [...selection];
        const deleteIndex = newSelection.indexOf(tag);
        newSelection.splice(deleteIndex, 1);
        setSelection(newSelection);
      } else {
        setValue(
          null as SDSAutocompleteValue<T, Multiple, DisableClearable, FreeSolo>
        );
        setSelection([]);
      }
    } else {
      handleTagDeleteMultiColumn(tag);
    }
  }
};
