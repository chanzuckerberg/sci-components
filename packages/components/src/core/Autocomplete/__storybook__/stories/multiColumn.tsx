import { Args } from "@storybook/react";
import { DefaultAutocompleteOption } from "src/core/Autocomplete/components/AutocompleteBase";
import { AUTOCOMPLETE_MULTI_COLUMN_OPTIONS } from "src/common/storybook/AUTOCOMPLETE_MULTI_COLUMN_OPTIONS";
import { useEffect, useState } from "react";
import { AutocompleteMultiColumnValue } from "src/core/Autocomplete";
import {
  AutocompleteChangeDetails,
  AutocompleteChangeReason,
  AutocompleteValue,
} from "@mui/base";
import {
  AUTOCOMPLETE_DIV_MARGIN,
  AUTOCOMPLETE_WRAPPER_STYLES,
} from "../constants";
import TagFilter from "src/core/TagFilter";
import RawAutocompleteMultiColumn from "src/core/Autocomplete/components/AutocompleteMultiColumn";

export const AutocompleteMultiColumnDemo = <
  T extends DefaultAutocompleteOption,
  Multiple extends boolean | undefined,
>(
  props: Args
): JSX.Element => {
  const {
    multiple,
    options = AUTOCOMPLETE_MULTI_COLUMN_OPTIONS,
    search,
    label,
  } = props;

  type DisableClearable = false;
  type FreeSolo = false;

  const [selection, setSelection] = useState<string[]>([]);

  const [theValue, setTheValue] =
    useState<
      AutocompleteMultiColumnValue<T, Multiple, DisableClearable, FreeSolo>
    >();

  useEffect(() => {
    setSelection([]);
  }, [multiple]);

  const handleChange = (
    _event: React.SyntheticEvent,
    newValue: AutocompleteMultiColumnValue<
      T,
      Multiple,
      DisableClearable,
      FreeSolo
    >,
    _reason: AutocompleteChangeReason,
    _details?: AutocompleteChangeDetails<T>
  ) => {
    setTheValue(newValue);
    const newSelection: string[] = [];

    if (multiple) {
      if (newValue) {
        Object.values(newValue).forEach((items) => {
          (items as T[])?.map(({ name }: { name: string }) => {
            newSelection.push(name);
          });
        });
        setSelection(newSelection);
      }
    } else {
      if (newValue) {
        Object.values(newValue).forEach((item) => {
          newSelection.push((item as T).name);
        });
      }
      setSelection(newSelection);
    }
  };

  return (
    <>
      <div style={AUTOCOMPLETE_WRAPPER_STYLES}>
        <RawAutocompleteMultiColumn<T, Multiple, DisableClearable, FreeSolo>
          label={label}
          disableCloseOnSelect={false}
          multiple={multiple}
          onChange={handleChange}
          options={options}
          search={search}
          value={theValue}
          {...props}
        />
      </div>
      <div style={{ margin: AUTOCOMPLETE_DIV_MARGIN }}>
        {selection.length
          ? selection.map((item) => {
              return (
                <TagFilter
                  key={item}
                  label={item}
                  onDelete={() => handleTagDelete(item)}
                  onClick={() => handleTagDelete(item)}
                />
              );
            })
          : null}
      </div>
    </>
  );

  function handleTagDelete(tag: string) {
    if (multiple) {
      if (theValue) {
        Object.entries(theValue).forEach(([key, values]) => {
          const index = (values as T[])?.findIndex(
            (item: T) => item.name === tag
          );
          if (index > -1) {
            const newValue = [...(values as T[])];
            newValue.splice(index, 1);
            setTheValue((prev) => {
              return {
                ...prev,
                [key]: newValue as AutocompleteValue<
                  T,
                  Multiple,
                  DisableClearable,
                  FreeSolo
                >,
              };
            });

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    } else {
      if (theValue) {
        Object.entries(theValue).forEach(([key, value]) => {
          if ((value as T).name === tag) {
            const finalValue = { ...theValue };
            delete finalValue[key];
            setTheValue(finalValue);

            const newSelection = [...selection];
            const deleteIndex = newSelection.indexOf(tag);
            newSelection.splice(deleteIndex, 1);
            setSelection(newSelection);
          }
        });
      }
    }
  }
};
