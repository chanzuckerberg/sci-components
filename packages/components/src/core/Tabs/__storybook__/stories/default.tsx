import { Args } from "@storybook/react";
import { useState } from "react";
import RawTabs, { Tab } from "src/core/Tabs";

export const Tabs = (props: Args): JSX.Element => {
  const { items, ...args } = props;

  const [value, setValue] = useState(0);

  const handleTabsChange = (_: React.SyntheticEvent, tabsValue: unknown) => {
    setValue(tabsValue as number);
  };

  return (
    <RawTabs {...args} value={value} onChange={handleTabsChange}>
      {items.map(({ count, label }: { count?: number; label: string }) => (
        <Tab label={label} count={count} key={label} />
      ))}
    </RawTabs>
  );
};
