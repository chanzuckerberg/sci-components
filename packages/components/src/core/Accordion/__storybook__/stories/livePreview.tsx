import { Args } from "@storybook/react";
import { Accordion } from "./default";

const livePreviewStyles = {
  display: "grid",
  gridColumnGap: "24px",
  gridRowGap: "50px",
  gridTemplateColumns: "repeat(4, 200px)",
};

export function LivePreviewDemo(props: Args): JSX.Element {
  return (
    <div style={livePreviewStyles as React.CSSProperties}>
      <Accordion
        id="accordion-1"
        togglePosition="right"
        useDivider
        {...props}
      />
      <Accordion
        id="accordion-2"
        togglePosition="right"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
      <Accordion id="accordion-3" togglePosition="left" useDivider {...props} />
      <Accordion
        id="accordion-4"
        togglePosition="left"
        subtitle="Optional Subtitle"
        useDivider
        {...props}
      />
    </div>
  );
}
