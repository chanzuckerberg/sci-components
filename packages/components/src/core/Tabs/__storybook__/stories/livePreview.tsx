import { Args } from "@storybook/react";
import { TABS_LIVE_PREVIEW_WRAPPER_STYLES } from "../constants";
import { Tabs } from "./default";
import { noop } from "src/common/utils";

export function LivePreviewDemo(props: Args): JSX.Element {
  const finalProps = {
    ...props,
    style: { width: "150px" },
  };

  return (
    <div style={TABS_LIVE_PREVIEW_WRAPPER_STYLES as React.CSSProperties}>
      <div>
        <Tabs
          onChange={noop}
          sdsSize="large"
          items={[{ label: "One" }, { label: "Two" }]}
          underlined
          {...finalProps}
        />
      </div>
      <div>
        <Tabs
          onChange={noop}
          sdsSize="large"
          items={[
            { count: 0, label: "One" },
            { count: 0, label: "Two" },
          ]}
          underlined
          {...finalProps}
        />
      </div>
      <div>
        <Tabs
          sdsSize="small"
          onChange={noop}
          items={[{ label: "One" }, { label: "Two" }]}
          underlined
          {...finalProps}
        />
      </div>
      <div>
        <Tabs
          sdsSize="small"
          onChange={noop}
          items={[
            { count: 0, label: "One" },
            { count: 0, label: "Two" },
          ]}
          underlined
          {...finalProps}
        />
      </div>
    </div>
  );
}
