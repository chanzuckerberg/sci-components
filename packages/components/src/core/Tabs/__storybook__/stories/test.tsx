import { Args } from "@storybook/react";
import { TABS_LIVE_PREVIEW_WRAPPER_STYLES } from "../constants";
import { Tabs } from "./default";
import { noop } from "src/common/utils";
import Tag from "src/core/Tag";
import { StyledTabCountWrapper } from "../style";

export function TestDemo(props: Args): JSX.Element {
  const finalProps = {
    ...props,
    "data-testid": "tabs",
    style: { width: "400px" },
  };

  return (
    <div>
      <div style={TABS_LIVE_PREVIEW_WRAPPER_STYLES as React.CSSProperties}>
        <div>
          <h4>Default</h4>
          <Tabs
            items={[{ count: 123, label: "Tab One" }, { label: "Tab Two" }]}
            onChange={noop}
            {...finalProps}
          />
        </div>
        <div>
          <h4>Small</h4>
          <Tabs
            items={[{ count: 123, label: "Tab One" }, { label: "Tab Two" }]}
            onChange={noop}
            sdsSize="small"
            {...finalProps}
          />
        </div>
        <div>
          <h4>Underlined</h4>
          <Tabs
            onChange={noop}
            items={[{ count: 123, label: "Tab One" }, { label: "Tab Two" }]}
            underlined
            {...finalProps}
          />
        </div>
      </div>
      <div style={TABS_LIVE_PREVIEW_WRAPPER_STYLES}>
        <div>
          <h4>Label with count</h4>
          <div>
            <Tabs
              onChange={noop}
              sdsSize="large"
              items={[
                { count: <BetaTagWithCount />, label: "One" },
                { label: "Two" },
              ]}
              underlined
              {...finalProps}
              style={{ width: "200px" }}
            />
          </div>
        </div>
      </div>
    </div>
  );
}

function BetaTagWithCount() {
  return (
    <>
      <StyledTabCountWrapper>123</StyledTabCountWrapper>
      <Tag
        label="BETA"
        hover={false}
        color="beta"
        sdsStyle="rounded"
        sdsType="secondary"
      />
    </>
  );
}
