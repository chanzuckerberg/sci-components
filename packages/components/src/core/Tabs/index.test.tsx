import { generateSnapshots } from "@chanzuckerberg/story-utils";
import { composeStory } from "@storybook/react";
import { render, screen } from "@testing-library/react";
import * as snapshotTestStoryFile from "./index.stories";
import Meta, { Test as TestStory } from "./index.stories";

// Returns a component that already contain all decorators from story level, meta level and global level.
const Test = composeStory(TestStory, Meta);

describe("<Tabs />", () => {
  generateSnapshots(snapshotTestStoryFile);

  it("renders Tabs component", () => {
    render(<Test />);

    const elements = screen.getAllByTestId("tabs");

    expect(elements.length).toBeTruthy();
  });

  it("renders tab texts", () => {
    render(<Test />);

    const Label1 = screen.getAllByText(/Tab One/i);
    const Label2 = screen.getAllByText(/Tab Two/i);

    expect(Label1.length).toBeTruthy();
    expect(Label2.length).toBeTruthy();
  });
});
