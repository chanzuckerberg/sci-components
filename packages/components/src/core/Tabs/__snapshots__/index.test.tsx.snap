// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> Default story renders snapshot 1`] = `
<div
  class="MuiTabs-root css-11fro2b-MuiTabs-root"
>
  <div
    class="MuiTabs-scroller MuiTabs-fixed css-jpln7h-MuiTabs-scroller"
    style="overflow: hidden; margin-bottom: 0px;"
  >
    <div
      class="MuiTabs-flexContainer css-heg063-MuiTabs-flexContainer"
      role="tablist"
    >
      <button
        aria-selected="true"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected css-17rzz1x-MuiButtonBase-root-MuiTab-root"
        role="tab"
        tabindex="0"
        type="button"
      >
        Upload from Your Computer
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-selected="false"
        class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary css-17rzz1x-MuiButtonBase-root-MuiTab-root"
        role="tab"
        tabindex="-1"
        type="button"
      >
        Upload from Basespace
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <span
      class="MuiTabs-indicator css-o80iit css-1aquho2-MuiTabs-indicator"
      style="left: 0px; width: 0px;"
    />
  </div>
</div>
`;
